MarsRover/Rover.cs

```diff
@@ -0,0 +1,68 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace MarsRover
+{
+    public class Rover
+    {
+        private Grid _grid;
+        private IHeading _heading;
+        private List<Obstacle> _obstacles;
+        private bool _obstacleDetected;
+
+        public Rover(Grid grid, IHeading heading, List<Obstacle> Obstacles)
+        {
+            _grid = grid.GridEntryAndSizeValidation(grid);
+            _heading = heading;
+            _obstacles = Obstacles;
+            _obstacleDetected = false;
+        }
+
+        public void TurnLeft() => _heading = _heading.TurnLeft();
+
+        public void TurnRight() => _heading = _heading.TurnRight();
+
+        public void MoveForward() => _grid = _heading.MoveForward(_grid);
+
+        public void MoveBackward() => _grid = _heading.MoveBackward(_grid);
+
+        public void InvalidCommand() { }
+
+        public string Direction() => _heading.GetHeading();
+
+        public int XCoordinate() => _grid.XCoordinate();
+
+        public int YCoodinate() => _grid.YCoordinate();
+
+        public int GridHeight() => _grid.GridHeight();
```

Comment:
`GridHeight` and `GridWidth` are not properties of a rover.  I would suggest removing these variables.  I'm not sure the rover should expose any properties of the `Grid` object from the `Rover` class.  I think doing this will greatly simplify some of the logic.